### ISSUE TITLE
UserDefaults subscribe problem

### ISSUE THREAD

**Issue by whisper-bye** [ReactiveX/RxSwift#1143]
Hi! I am new here about rxswift, I have some issue about the UserDefaults.standard.rx

I set this at my viewDidLoad method
```
UserDefaults.standard.rx.observe(String.self, "currentTileOverlay")
            .subscribe(onNext: { (value) in
            ...
```

then I set this at my cell tableView.rx.modelSelected
```swift
UserDefaults.standard.set...
```

when I clicked a cell, but the subscribe method submit twice
so what should I do to fix it?
thx a lot!

**RxSwift/RxCocoa/RxBlocking/RxTest version/commit**

  _version or commit here_

**Platform/Environment**

  - [x] iOS
  - [ ] macOS
  - [ ] tvOS
  - [ ] watchOS
  - [ ] playgrounds

**How easy is to reproduce? (chances of successful reproduce after running the self contained code)**

  - [x] easy, 100% repro
  - [ ] sometimes, 10%-100%
  - [ ] hard, 2% - 10%
  - [ ] extremely hard, %0 - 2%

**Xcode version**:

```
  Xcode latest release
```

:warning: Fields below are optional for general issues or in case those questions aren't related to your issue, but filling them out will increase the chances of getting your issue resolved. :warning:

**Installation method**:
  - [x] CocoaPods
  - [ ] Carthage
  - [ ] Git submodules

**I have multiple versions of Xcode installed**:
  (so we can know if this is a potential cause of your issue)
  - [ ] yes (which ones)
  - [x] no

**Level of RxSwift knowledge**:
  (this is so we can understand your level of knowledge
    and formulate the response in an appropriate manner)
  - [ ] just starting
  - [ ] I have a small code base
  - [x] I have a significant code base

**Comment by kzaher** [comment]
Hi @whisper-bye ,

please take a look at RxExample how `disposeBag` are used.

**Comment by whisper-bye** [comment]
@kzaher 
yes I did, but there is nothing about disposeBag

**Comment by whisper-bye** [comment]
@kzaher  my simple test code

```swift
import UIKit

import RxSwift
import RxCocoa
import RxDataSources

class ViewController: UIViewController {
    let disposeBag = DisposeBag()
    
    @IBOutlet weak var tableView: UITableView! {
        didSet {
            tableView.register(UITableViewCell.self, forCellReuseIdentifier: String(describing: UITableViewCell.self))
            
            tableView.rx
                .itemSelected
                .subscribe { (indexPath) in
                    UserDefaults.standard.set("\(indexPath)", forKey: "key")
                }
                .disposed(by: disposeBag)
        }
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        // Do any additional setup after loading the view, typically from a nib.
        UserDefaults.standard.rx
            .observe(String.self, "key")
            .subscribe(onNext: { (value) in
                if let value = value {
                    print(value)
                }
            })
            .disposed(by: disposeBag)
        
        let dataSource = RxTableViewSectionedReloadDataSource<SectionModel<String, String>>()
        
        dataSource.configureCell = { (dataSource, tableView, indexPath, item) in
            let cell = tableView.dequeueReusableCell(withIdentifier: String(describing: UITableViewCell.self), for: indexPath)
            cell.textLabel?.text = item
            
            return cell
        }
        
        Observable.just([SectionModel(model: "", items: (0..<5).map({ "\($0)" }))])
            .bindTo(tableView.rx.items(dataSource: dataSource))
            .disposed(by: disposeBag)
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        
        // Dispose of any resources that can be recreated.
    }
}

```

**Comment by kzaher** [comment]
Hi @whisper-bye ,

I don't understand the question.

**Comment by whisper-bye** [comment]
@kzaher 
my question is, when I clicked a cell, but the subscribe method submit twice? It should be once.
strange debug print

2017-03-18 22:00:39.008: ViewController.swift:28 (viewDidLoad()) -> subscribed
2017-03-18 22:00:39.183: ViewController.swift:28 (viewDidLoad()) -> Event next(Optional("next([0, 0])"))
next([0, 0])
2017-03-18 23:22:47.962: ViewController.swift:28 (viewDidLoad()) -> Event next(Optional("next([0, 4])"))
next([0, 4])
2017-03-18 23:22:48.038: ViewController.swift:28 (viewDidLoad()) -> Event next(Optional("next([0, 4])"))
next([0, 4])
2017-03-18 23:22:51.047: ViewController.swift:28 (viewDidLoad()) -> Event next(Optional("next([0, 1])"))
next([0, 1])
2017-03-18 23:22:51.050: ViewController.swift:28 (viewDidLoad()) -> Event next(Optional("next([0, 1])"))
next([0, 1])

**Comment by kzaher** [comment]
Hi @whisper-bye ,

I don't think this is a RxSwift bug. Here is Stack Overflow post

http://stackoverflow.com/questions/36193503/kvo-broken-in-ios-9-3

and here is a proof that KVO works properly.

```swift
class ViewController: UIViewController {
    let disposeBag = DisposeBag()

    let defaults = UserDefaults.standard

    dynamic var key: String? = nil
    
    @IBOutlet weak var tableView: UITableView! {
        didSet {

            let defaults = self.defaults
            tableView.register(UITableViewCell.self, forCellReuseIdentifier: String(describing: UITableViewCell.self))

            tableView.rx
                .itemSelected
                .observeOn(MainScheduler.asyncInstance)
                .subscribe { [weak self] (indexPath) in
                    let value: NSString = NSString(string: "\(indexPath)")
                    print("setting \(value)")
                    self?.key = String(value)
                    defaults.set(value, forKey: "key")
                }
                .disposed(by: disposeBag)
        }
    }

    override func viewDidLoad() {
        super.viewDidLoad()

        // Do any additional setup after loading the view, typically from a nib.
        self.rx
            .observe(String.self, "key")
            .debug("a")
            .subscribe(onNext: { (value) in
                print(value)
            })
            .disposed(by: disposeBag)

        defaults.rx
            .observe(String.self, "key")
            .debug("b")
            .subscribe(onNext: { (value) in
                print(value)
            })
            .disposed(by: disposeBag)

        let dataSource = RxTableViewSectionedReloadDataSource<SectionModel<String, String>>()

        dataSource.configureCell = { (dataSource, tableView, indexPath, item) in
            let cell = tableView.dequeueReusableCell(withIdentifier: String(describing: UITableViewCell.self), for: indexPath)
            cell.textLabel?.text = item

            return cell
        }

        Observable.just([SectionModel(model: "", items: (0..<5).map({ "\($0)" }))])
            .bindTo(tableView.rx.items(dataSource: dataSource))
            .disposed(by: disposeBag)
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()

        // Dispose of any resources that can be recreated.
    }
}
```

**Comment by whisper-bye** [comment]
@kzaher 
thanks for your reply, it's very helpful, I will investigate and report it later

**Comment by whisper-bye** [comment]
I wrote a simple test by using raw KVO API
```swift
UserDefaults.standard.addObserver(self, forKeyPath: "key", options: [.new], context: nil)
// and then listen on
override func observeValue(forKeyPath keyPath: String?, of object: Any?, change: [NSKeyValueChangeKey : Any]?, context: UnsafeMutableRawPointer?) {
    print(keyPath!, object!, change!, context!)
}
```
the same result, observeValue method submit twice
```
key <NSUserDefaults: 0x6080000416b0> [__C.NSKeyValueChangeKey(_rawValue: new): [0, 3], __C.NSKeyValueChangeKey(_rawValue: kind): 1] 0x00007fd166d03c68
key <NSUserDefaults: 0x6080000416b0> [__C.NSKeyValueChangeKey(_rawValue: new): [0, 3], __C.NSKeyValueChangeKey(_rawValue: kind): 1] 0x00007fd166d03c68
key <NSUserDefaults: 0x6080000416b0> [__C.NSKeyValueChangeKey(_rawValue: new): [0, 1], __C.NSKeyValueChangeKey(_rawValue: kind): 1] 0x00007fd166d03c68
key <NSUserDefaults: 0x6080000416b0> [__C.NSKeyValueChangeKey(_rawValue: new): [0, 1], __C.NSKeyValueChangeKey(_rawValue: kind): 1] 0x00007fd166d03c68
```
then I found these https://forums.developer.apple.com/thread/47609,
according to the Apple Staff's say, it's fixed in 9.3.2, but my target env is 10.2, unfortunately, maybe not.

a quick workaround, it is works fine with me
```swift
UserDefaults.standard.rx
    .observe(String.self, "key")
+   .debounce(0.1, scheduler: MainScheduler.asyncInstance)
    ...
```

**Comment by nmdias** [comment]
I have this bug on iOS 10.3. Calling `distinctUntilChanged` as a workaround did it for me.
